docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,608 @@
+It should also help plugin/extension developers to define their topic schemes as well, inline with the tedge topics.
+
+# Background
+
+The inconsistency in the existing topic schemes of thin-edge has long been a problem for both users and the
+thin-edge dev team to write new applications or new extensions of thin-edge.
+
+Here are a few such examples:
+1. Topic for events: `tedge/events/{event-type}/{child-id}`
+1. Topic for firmware update commands: `tedge/{child-id}/commands/req/firmware_update`
+1. Topic for software update commands: `tedge/commands/req/software/list`
+
+... where there is inconsistency in the placement of child devices and how the commands are grouped.
+
+There are a few other limitations like the lack of support for services on the thin-edge device,
+difficulty in extending existing topics with additional sub-topics, etc
+which are detailed in the requirements section.
+
+# Domain model influence
+
+The MQTT topics and interactions are modelled around the following entities:
+
+1. **Device**
+   A device can be viewed as a composition of hardware components and firmware/software components running on it.
+   The device can extract data from these hardware components and emit that as telemetry data using some software on it.
+   The device can also control these hardware components using some software.
+   The device also manages the firmware and all the software running on it.
+   A device could be connected to many other devices as well as multiple cloud providers/instances.
+1. **Tedge device**
+   The gateway device that is connected to the cloud, where thin-edge.io is installed,
+   which emits its own telemetry data and can receive commands to perform operations on the device.
+2. **Child device**
+   Typically, a different device that is connected to the tedge device, which has its own identity
+   which is separate from the tedge device itself.
+   It also emits its own measurements and can receive commands.
+   A child device need not always be a a physical device, but can also be a logical one,
+   abstracting some parts or groups of parts of the main device as well, with its own identity.
+   A child device can have further nested child devices as well.
+3. **Service**
+   A service can be a component or a piece of software running on a device (tedge device or child device).
+   For example, a device can have a cloud connector/agent software that can be viewed as a service.
+   Any software on the device can be modelled as a service, if monitoring them separately from the device makes sense.
+   This abstraction be also be used to isolate various aspects of the device itself into separate groups, but still linked to the device.
+   For example, if you model a car as a device, the sensor data coming from different tyres, breaks, engine, infotainment system etc
+   can be isolated into their own respective services (more like light-weight child devices).
```

Comment:
In most cases, yes, that's the most natural way to visualise/model a deployment.

But, if the hardware or a collection of such hardware is very limited and can only generate telemetry data and need not/can not handle any operations, a user might choose to model those as light-weight services instead of a heavy-weight child device, primarily for cost optimisation, as many cloud providers charge extra for each extra device.

Reply:
> as many cloud providers charge extra for each extra device.

Okay - but this highlights that distinguishing child-device vs service is of little significance.

Reply:
To avoid confusion we could remove this specific example. Yes technically anyone can model any component as a child device or service, but we should try to guide the users to the best practices / typical setups. We could have an additional note about some advanced setups where an entity could be modelled as a service (if needed).